drivers: i2c_nrfx_twim: add exclusive access API

ankuns opened the pull request:
This commit provides an extension to the i2c_nrfx_twim driver. It introduces possibility to acquire/release exclusive access to the i2c bus.

**Background:**
The need for such feature comes from nRF Connect SDK, where i2c write transfers (on nRF SoCs only) must be performed while having a timeslot granted by a scheduler external to Zephyr. The scheduler has very strict timing requirements, which cannot be met if i2c transfers are performed from a thread context (even of the highest priority). The solution is to make i2c transfers asynchronous relying only on underlying nrfx_twim driver from ZLI priority. To make this approach work with regular i2c.h - compatible drivers exclusive access to the i2c bus is necessary.

**Usage:**
From a thread context call `i2c_nrfx_twim_exclusive_access_acquire`
Then perform nrfx_twim direct access calls of your choice, possibly asynchronous transfers.
From a thread context call `i2c_nrfx_twim_exclusive_access_release`

teburd commented:
This problem was solved in a general way with .submit already using lock-free queues, is there some reason we need to add a custom API here?

bjarki-andreasen commented:
@teburd Well, @JordanYates added RTIO support to the TWIM, I added an RTIO I2C sample an a few patches :) 

@ankuns See https://github.com/zephyrproject-rtos/zephyr/tree/main/samples/drivers/i2c/rtio_loopback Is there a reason not to use RTIO?

ankuns commented:
@bjarki-andreasen , does the I2C RTIO support operation on ZLI priority?

teburd commented:
rtio is completely detached from the scheduler, if you want you can start interrupts from any context, including zli's, the entire thing is built on a lock-free queue of I/O like work given to a driver. This can be done at any time, from any where. Safely!

bjarki-andreasen commented:
> @bjarki-andreasen , does the I2C RTIO support operation on ZLI priority? My use case is that I need to perform I2C transfers in a time window given by a scheduler external to Zephyr. I get the notification from the scheduler in ZLI interrupt. The scheduler does not tolerate any delays.

Not tolerating delays on a shared bus is does not seem to be a valid requirement, what if there is already a transaction in progress? In that case you will still need to wait for the transaction to be complete before aquiring the bus. Seems like the solution is to not share the bus in the first place :)

nordic-piks commented:
Please add tests/samples with use cases you have described.
(or extend existing ones)

ankuns commented:
> Not tolerating delays on a shared bus does not seem to be a valid requirement, what if there is already a transaction in progress? In that case you will still need to wait for the transaction to be complete before aquiring the bus. Seems like the solution is to not share the bus in the first place :)

@bjarki-andreasen I call the `i2c_nrfx_twim_exclusive_access_acquire` when I need to start my operation. If there is a transaction on the bus this call will block until it is finish, then underlying semaphore is taken. This delay can be tolerated. When I have the exclusive access (underlying semaphore taken) I request to the external scheduler for a timeslot. I expect it will give me the timeslot in around 0...tens of milliseconds. But when it gives me a timeslot I must perform my transfer immediately, with no delays. After the transfer I call `i2c_nrfx_twim_exclusive_access_release` so that other threads can get the underlying access semaphore.

bjarki-andreasen commented:
@teburd Bad question, can we introduce a lock to RTIO :D Maybe we could have a `RTIO_OP_AWAIT` which simply stops the current transaction, until some event triggers it, a "delay" type operation :)

teburd commented:
> > rtio is completely detached from thread scheduler with two ways to opt into getting threads to wait for completions if needed. You can start interrupts from any context, including ZLI interrupts. The entire thing is built on a lock-free queue of I/O like work given to a driver. This can be done at any time, from any where. Safely!
> 
> Are you certain it is safe from ZLIs? zephyr APIs can not be used safely from ZLIs, since the kernel has no way of synchronizing access to data, given ZLIs can not be masked by the kernel (k_spin_lock, k_sem, k_mutex etc. are not safe, they don't prevent ZLIs from accessing the same resource). Hacks like
> 
> https://github.com/zephyrproject-rtos/zephyr/blob/fb4bb5d7688f5b232023ea07308edfe734f41b7e/drivers/clock_control/clock_control_nrf2_hfxo.c#L81-L96
> are needed to allow ZLIs to directly call anything zephyr, except for the atomic API and const read-only access.

Yes, I'm positive, though you would need to poll the completion queue rather than using a semaphore.

So

CONFIG_RTIO_CONSUME_SEM=n
CONFIG_RTIO_SUBMIT_SEM=n

In a thread waiting completion...

```c
rtio_submit(r, 0); /* Submit asynchronous work to do, can be done from *anywhere* including a ZLI */ 
```

```c
/* In a thread somewhere... wait on a completion */
struct rtio_cqe *cqe;
do {
   cqe = rtio_cqe_consume(r);
} while (cqe == NULL);
```

This would work just fine, zero kernel calls are made.

teburd commented:
> @teburd Bad question, can we introduce a lock to RTIO :D Maybe we could have a `RTIO_OP_AWAIT` which simply stops the current transaction, waiting until some event occurs, a "delay" type operation :)
> 
> An initial proposal:
> 
> ```
> struct rtio_sqe {
>         ...
> 		/** OP_AWAIT */
> 		struct {
> 			bool signaled; /**< set by calling rtio_sqe_signal(struct rtio_sqe *sqe) */
> 		} await;
>         ...
> ```
> 
> ```
> void rtio_sqe_signal(struct rtio_sqe *sqe)
> {
>         sqe->await.signaled = true;
>         /* check if iodev is awaiting this */
>         /* submit next sqe if yes */
> }
> ```
> 
> if the sqe when submitted is already signaled, its is treated as a nop :)

Exclusive access is not something built into the current APIs. We could add it to RTIO, and it would need to be asynchronous acquire/release kind of pairing to work correctly because there may be asynchronous transactions already in flight still.

So to do that we'd want basically a RTIO_SQE_ACQUIRE_OP and RTIO_SQE_RELEASE_OP pair, which could return an exclusively owned iodev the user could submit work to. Only this exclusive queue would be pulled from while its checked out.

ankuns committed:
drivers: i2c_nrfx_twim: add exclusive access API

This commit provides an extension to the i2c_nrfx_twim driver.
It introduces possibility to acquire/release exclusive access to the
i2c bus. While the exclusive access to the i2c bus is acquired
you can access the underlying nrfx_twim driver directly.

Signed-off-by: Andrzej Kuros <andrzej.kuros@nordicsemi.no>

ankuns commented:
@carlescufi @anangl @bjarki-andreasen @nika-nordic according to the internal discussion the PR is changed so that only the `i2c_nrfx_twim_exclusive_access_acquire` and `i2c_nrfx_twim_exclusive_access_release` stay in Zephyr.

teburd commented:
Reassigned as it isn't really a question of i2c anymore

bjarki-andreasen commented:
ping @anangl
